jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/LocalBaseImageSteps.java

```diff
@@ -88,27 +89,67 @@ static boolean isGzipped(Path path) throws IOException {
     }
   }
 
-  private final ExecutorService executorService;
-  private final BuildConfiguration buildConfiguration;
-  private final Path tarPath;
-  private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
-  private final TempDirectoryProvider tempDirectoryProvider;
+  static Callable<LocalImage> retrieveDockerDaemonImageStep(
+      ExecutorService executorService,
+      BuildConfiguration buildConfiguration,
+      ProgressEventDispatcher.Factory progressEventDispatcherFactory,
+      DockerClient dockerClient) {
+    return () -> {
+      ImageReference imageReference = buildConfiguration.getBaseImageConfiguration().getImage();
+      try (TempDirectoryProvider tempDirectoryProvider = new TempDirectoryProvider();
+          ProgressEventDispatcher progressEventDispatcher =
+              progressEventDispatcherFactory.create("processing base image " + imageReference, 2)) {
+        Path outputDir = tempDirectoryProvider.newDirectory();
+        Path tarPath = outputDir.resolve("out.tar");
+        try (TimerEventDispatcher ignored =
+            new TimerEventDispatcher(
+                buildConfiguration.getEventHandlers(),
+                "Saving " + imageReference + " from Docker daemon")) {
+          long size = dockerClient.inspect(imageReference).getSize();
+          try (ProgressEventDispatcher dockerProgress =
+                  progressEventDispatcher
+                      .newChildProducer()
+                      .create("saving base image " + imageReference, size);
+              ThrottledAccumulatingConsumer throttledProgressReporter =
+                  new ThrottledAccumulatingConsumer(dockerProgress::dispatchProgress)) {
+            dockerClient.save(imageReference, tarPath, throttledProgressReporter);
+          }
+        }
 
-  ExtractTarStep(
+        return cacheDockerImageTar(
+            buildConfiguration,
+            executorService,
+            tarPath,
+            tempDirectoryProvider,
+            progressEventDispatcher.newChildProducer());
+      }
+    };
+  }
+
+  static Callable<LocalImage> retrieveTarImageStep(
       ExecutorService executorService,
       BuildConfiguration buildConfiguration,
-      Path tarPath,
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
-      TempDirectoryProvider tempDirectoryProvider) {
-    this.executorService = executorService;
-    this.buildConfiguration = buildConfiguration;
-    this.tarPath = tarPath;
-    this.progressEventDispatcherFactory = progressEventDispatcherFactory;
-    this.tempDirectoryProvider = tempDirectoryProvider;
+      Path tarPath) {
+    return () -> {
+      try (TempDirectoryProvider tempDirectoryProvider = new TempDirectoryProvider()) {
```

Comment:
Now it seems like you don't need to pass `tempDirectoryProvider`, because there's no more chaining between `Step`s. I know `retrieveDockerDaemonImageStep` has to use a temp dir provider, but I still think `cacheDockerImageTar` doesn't need to require passing one. It can create one inside it.